src/cmd/ksh93/sh/subshell.c

```diff
@@ -503,9 +503,14 @@ Sfio_t *sh_subshell(Shell_t *shp, Shnode_t *t, volatile int flags, int comsub) {
         // Save trap table.
         shp->st.otrapcom = 0;
         shp->st.otrap = savst.trap;
-        if ((nsig = shp->st.trapmax * sizeof(char *)) > 0 || shp->st.trapcom[0]) {
-            nsig += sizeof(char *);
-            memcpy(savsig = malloc(nsig), (char *)&shp->st.trapcom[0], nsig);
+        nsig = shp->st.trapmax;
+        if (nsig > 0 || shp->st.trapcom[0]) {
```

Comment:
I know you're just copying the original statement but now is the time to think about whether it is correct. Shouldn't it be impossible for `st.trapmax` to be zero yet a trap be defined? Shouldn't this just be `if (nsig)`?

Reply:
I think that's why `shp->st.trapcom[0]` is being checked here.

Reply:
I admit I haven't looked closely at the code to understand how `st.trapmax` is used. If it is the maximum offset into the `st.trapcom` array then it should be set to -1 rather than zero if the array is empty. In which case the check of `st.trapcom[0]` isn't needed. Regardless, it would be better would be to redefine it as `st.trapcount` so that zero means there are no traps.

In any case ignore this rant. I'm just pointing out how stupid and confusing the existing code is about such matters. The existing pattern pretty much guarantees bugs will be present in any code that manipulates shell traps. Which was why I was suggesting expanding the scope of the change to fix the core problem.